components/tools/ToolDashboard.js

```diff
@@ -0,0 +1,155 @@
+    const [checkedCategory, setCheckedCategory] = useState(categories)
+    const [checkToolsList, setcheckToolsList] = useState(true)
+
+    const handleApplyCategory = () => {
+        setCategories(checkedCategory)
+        setopenFilter({
+            filter: false,
+            category: false
+        })
+    }
+    useEffect(() => {
+        const checkIfClickOutside = (e) => {
+            if((openFilter.filter && filterRef.current && !filterRef.current.contains(e.target)) || openFilter.category && categoryRef.current && !categoryRef.current.contains(e.target))
+                setopenFilter({
+                    filter: false,
+                    category: false
+                })             
+        }
+        document.addEventListener("mousedown", checkIfClickOutside)
+        return () => {
+            document.removeEventListener("mousedown", checkIfClickOutside)
+        }
+    })
+
+    useEffect(()=> {
+        let tempToolsList = {}
+        if(categories.length >0){
+            for(let category of categories){
+                Object.keys(ToolsData).forEach((key) => {
+                    if(key===category) tempToolsList[key] = JSON.parse(JSON.stringify(ToolsData[key]))
+                })
+            }
+        }else{
+            tempToolsList = JSON.parse(JSON.stringify(ToolsData));
+        }
+        setcheckToolsList(false)
+        Object.keys(tempToolsList).forEach((category) => {
+            tempToolsList[category].toolsList = tempToolsList[category].toolsList.filter((tool) => {
+                let isLanguageTool = true, isTechnologyTool = true, isSearchTool = true, isAsyncAPITool = true;
+                if(languages.length){
+                    isLanguageTool = false;
+                    for(let language of languages){
+                        if(language === tool.filters.language.name) isLanguageTool = true;
+                    }
+                }
+                if(technologies.length){
+                    isTechnologyTool = false;
+                    for(let technology of technologies){
+                        if(tool.filters.technology.find((item) => item.name === technology)) isTechnologyTool = true;
+                    }
+                }
+                if(searchName){
+                    isSearchTool = tool.title.toLowerCase().includes(searchName.toLowerCase())
+                }
+                if(isAsyncAPIOwner)
+                    isAsyncAPITool = tool.filters.isAsyncAPIOwner === isAsyncAPIOwner ? true : false
+                    if(category === 'IDE Extensions'){
+                        console.log(tool.filters.hasCommercial, isPaid)
+                    }
+                return isLanguageTool && isTechnologyTool && isSearchTool && isAsyncAPITool && tool.filters.hasCommercial === isPaid;
+            })
+            if(tempToolsList[category].toolsList.length) setcheckToolsList(true)
+        })
+        setToolsList(tempToolsList)
```

Comment:
Probably you should put that logic to separate function 😅

Reply:
Done 👍